File: leetcode/1213-intersection-of-three-sorted-arrays/intersection-of-three-sorted-arrays.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @param {number[]} arr3
 * @return {number[]}
 */
var arraysIntersection = function(arr1, arr2, arr3) {
  const result = [];

  let it1 = 0;
  let it2 = 0;
  let it3 = 0;

  while (
    it1 < arr1.length &&
    it2 < arr2.length &&
    it3 < arr3.length
    ) {
    const maxOfThree = Math.max(arr1[it1], arr2[it2], arr3[it3]);

    while(arr1[it1] < maxOfThree) {
      it1++;
    }

    while(arr2[it2] < maxOfThree) {
      it2++;
    }

    while(arr3[it3] < maxOfThree) {
      it3++;
    }

    if (arr1[it1] === arr2[it2] && arr2[it2] === arr3[it3]) {
      result.push(maxOfThree);
    }

    if (arr1[it1] === maxOfThree) it1++;
    if (arr2[it2] === maxOfThree) it2++;
    if (arr3[it3] === maxOfThree) it3++;
  }

  return result;
};

File: leetcode/1213-intersection-of-three-sorted-arrays/IntersectionOfThreeSortedArrays.java
class Solution {
  public List<Integer> arraysIntersection(int[] arr1, int[] arr2, int[] arr3) {
    List<Integer> result = new ArrayList<>();

    for (int i = 0, j = 0, k = 0; i < arr1.length && j < arr2.length && k < arr3.length;) {
      if (arr1[i] == arr2[j] && arr2[j] == arr3[k]) {
        result.add(arr1[i]);
        i++; j++; k++;
      } else {
        int maxAcross = Math.max(arr1[i], Math.max(arr2[j], arr3[k]));

        if (arr1[i] < maxAcross) {
          i++;
        }

        if (arr2[j] < maxAcross) {
          j++;
        }

        if (arr3[k] < maxAcross) {
          k++;
        }
      }
    }

    return result;
  }
}
